packages/auth/src/providers/cognito/apis/signInWithRedirect.ts

```diff
@@ -0,0 +1,454 @@
+	const base64URLFromAWSCrypto = _base64URL(b64);
+
+	return base64URLFromAWSCrypto;
+}
+
+function _base64URL(string) {
+	return string.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
+}
+
+function _generateRandom(size: number) {
+	``;
+	const CHARSET =
+		'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+	const buffer = new Uint8Array(size);
+	if (typeof window !== 'undefined' && !!window.crypto) {
+		window.crypto.getRandomValues(buffer);
+	} else {
+		for (let i = 0; i < size; i += 1) {
+			buffer[i] = (Math.random() * CHARSET.length) | 0;
+		}
+	}
+	return _bufferToString(buffer);
+}
+
+function _bufferToString(buffer: Uint8Array) {
+	const CHARSET =
+		'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+	const state = [];
+	for (let i = 0; i < buffer.byteLength; i += 1) {
+		const index = buffer[i] % CHARSET.length;
+		state.push(CHARSET[index]);
+	}
+	return state.join('');
+}
+
+async function handleCodeFlow({
+	currentUrl,
+	userAgentValue,
+	clientId,
+	redirectUri,
+	domain,
+}: {
+	currentUrl: string;
+	userAgentValue?: string;
+	clientId: string;
+	redirectUri: string;
+	domain: string;
+}) {
+	/* Convert URL into an object with parameters as keys
+{ redirect_uri: 'http://localhost:3000/', response_type: 'code', ...} */
+	const url = new URL(currentUrl);
+	try {
+		_validateStateFromURL(url);
+	} catch (err) {
+		resolveInflight();
+
+		resolveInflight = () => {};
+		// clear temp values
+		window.localStorage.removeItem('com.amplify.cognito.pkce');
+		window.localStorage.removeItem('com.amplify.cognito.state');
+		window.localStorage.removeItem('com.amplify.cognito.inflightOAuth');
+		return;
+	}
+	const code = url.searchParams.get('code');
+
+	const currentUrlPathname = url.pathname || '/';
+	const redirectUriPathname = new URL(redirectUri).pathname || '/';
+
+	if (!code || currentUrlPathname !== redirectUriPathname) {
+		return;
+	}
+
+	const oAuthTokenEndpoint = 'https://' + domain + '/oauth2/token';
+
+	// TODO(v6): check hub events
+	// dispatchAuthEvent(
+	// 	'codeFlow',
+	// 	{},
+	// 	`Retrieving tokens from ${oAuthTokenEndpoint}`
+	// );
+
+	const code_verifier = window.localStorage.getItem('com.amplify.cognito.pkce');
+
+	const oAuthTokenBody = {
+		grant_type: 'authorization_code',
+		code,
+		client_id: clientId,
+		redirect_uri: redirectUri,
+		...(code_verifier ? { code_verifier } : {}),
+	};
+
+	const body = Object.entries(oAuthTokenBody)
+		.map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
+		.join('&');
+
+	const {
+		access_token,
+		refresh_token,
+		id_token,
+		error,
+		token_type,
+		expires_in,
+	} = await (
+		(await fetch(oAuthTokenEndpoint, {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded',
+				[USER_AGENT_HEADER]: userAgentValue,
+			},
+			body,
+		})) as any
+	).json();
+
+	if (error) {
+		resolveInflight();
+
+		resolveInflight = () => {};
+		throw new Error(error);
```

Comment:
Should be mapped as  AuthError
```suggestion
       throw new AuthError({
       name:'OAuthException', 
       message:'error while trying to finish the OAuth flow', 
       underlayingException:error
       })
```